File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofStores/NBABettingTrendsStore.swift
import Foundation
import Observation
import Supabase
import WagerproofModels
import WagerproofServices

/// NBA situational betting trends store. Mirrors RN
/// `hooks/useNBABettingTrends.ts` end-to-end plus the cached
/// `useNBABettingTrendsForGame` helper. Fetches from
/// `nba_game_situational_trends_today`; falls back to
/// `nba_game_situational_trends` when the today view is empty (matches RN).
/// After loading the rows it pairs them by `team_side`, joins tipoff times
/// from `nba_input_values_view`, and computes the O/U consensus + ATS
/// dominance scores used for sorting.
///
/// Backend queries are byte-identical to RN — the select clauses, table
/// names, and post-fetch joins must NOT diverge. Aligned with sibling
/// `NCAABBettingTrendsStore` (see ticket #070 for parity between sports).
@Observable
@MainActor
public final class NBABettingTrendsStore {
    public enum SortMode: String, Hashable, Sendable {
        case time, ouConsensus, atsDominance
    }

    public enum LoadState: Equatable, Sendable {
        case idle, loading, loaded, refreshing
        case failed(String)
    }

    public private(set) var games: [NBAGameTrendsData] = []
    public private(set) var loadState: LoadState = .idle
    /// Set on every successful refresh (including the DummyData branch).
    /// Drives both the TTL guard below and the game sheets' first-hydrate
    /// skeleton rule (`loading && lastFetched == nil`).
    public private(set) var lastFetched: Date?

    private let minGamesThreshold: Int = 5
    private let minPercentage: Double = 55
    private let minATSDifference: Double = 10

    public init() {}

    /// Lookup a single game by id. Mirrors RN's
    /// `useNBABettingTrendsForGame(gameId)` which reads from the same
    /// module-level cache.
    public func trends(for gameId: Int) -> NBAGameTrendsData? {
        games.first(where: { $0.gameId == gameId })
    }

    /// Idempotent hydrate for sheet-local stores — skips the network while a
    /// fetch is in flight or the slate is fresh, so swiping through the game
    /// carousel doesn't refetch the whole trends view per page. Mirrors
    /// `MLBBettingTrendsStore.refreshIfNeeded`.
    public func refreshIfNeeded(maxAge: TimeInterval = 600) async {
        if loadState == .loading || loadState == .refreshing { return }
        if let lastFetched, Date().timeIntervalSince(lastFetched) < maxAge { return }
        await refresh()
    }

    public func refresh() async {
        #if DEBUG
        // Dummy Data Mode: synthesized situational trends keyed to the captured
        // NBA slate so the per-game betting-trends widget populates offseason.
        if DummyDataMode.isEnabled {
            games = sortGames(DummyData.nbaTrendsData(), mode: .time)
            loadState = .loaded
            lastFetched = Date()
            return
        }
        #endif
        loadState = .loading
        let cfb = await CFBSupabase.shared.client

        // Step 1: situational trends (today's table → fallback).
        var trendsRows: [NBASituationalTrendRow] = []
        do {
            let primary: [NBASituationalTrendRow] = (try? await cfb
                .from("nba_game_situational_trends_today")
                .select()
                .order("game_date", ascending: true)
                .order("game_id", ascending: true)
                .execute()
                .value) ?? []
            if !primary.isEmpty {
                trendsRows = primary
            } else {
                // Fallback path mirrors RN.
                let fallback: [NBASituationalTrendRow] = try await cfb
                    .from("nba_game_situational_trends")
                    .select()
                    .order("game_date", ascending: true)
                    .order("game_id", ascending: true)
                    .execute()
                    .value
                trendsRows = fallback
            }
        } catch {
            loadState = .failed("Failed to fetch NBA trends: \(error.localizedDescription)")
            return
        }
        if trendsRows.isEmpty {
            games = []
            loadState = .loaded
            lastFetched = Date()
            return
        }

        // Step 2: group rows by game_id using team_side.
        var partial: [Int: (away: NBASituationalTrendRow?, home: NBASituationalTrendRow?, gameDate: String)] = [:]
        for row in trendsRows {
            let side = row.teamSide.lowercased()
            guard side == "away" || side == "home" else { continue }
            var existing = partial[row.gameId] ?? (nil, nil, row.gameDate)
            if side == "away" {
                existing.away = row
            } else {
                existing.home = row
            }
            partial[row.gameId] = existing
        }

        var built: [NBAGameTrendsData] = []
        for (gameId, slots) in partial {
            guard let away = slots.away, let home = slots.home else { continue }
            built.append(
                NBAGameTrendsData(
                    gameId: gameId,
                    gameDate: slots.gameDate,
                    tipoffTime: nil,
                    awayTeam: away,
                    homeTeam: home
                )
            )
        }

        // Step 3: tipoff times from nba_input_values_view for today's game ids.
        if !built.isEmpty {
            let gameIds = built.map { $0.gameId }
            let timeRows: [TipoffRow] = (try? await cfb
                .from("nba_input_values_view")
                .select("game_id, tipoff_time_et")
                .in("game_id", values: gameIds)
                .execute()
                .value) ?? []
            var timesMap: [Int: String] = [:]
            for row in timeRows {
                if let t = row.tipoffTimeEt { timesMap[row.gameId] = t }
            }
            built = built.map { game in
                var copy = game
                copy.tipoffTime = timesMap[game.gameId]
                return copy
            }
        }

        games = sortGames(built, mode: .time)
        loadState = .loaded
        lastFetched = Date()
    }

    // MARK: - Sort scoring (mirrors RN logic byte-for-byte)

    private func ouConsensusStrength(for game: NBAGameTrendsData) -> Double {
        var total: Double = 0
        let buckets: [(awayOver: Double?, awayUnder: Double?, awayRec: String?,
                       homeOver: Double?, homeUnder: Double?, homeRec: String?)] = [
            (game.awayTeam.ouLastGameOverPct, game.awayTeam.ouLastGameUnderPct, game.awayTeam.ouLastGameRecord,
             game.homeTeam.ouLastGameOverPct, game.homeTeam.ouLastGameUnderPct, game.homeTeam.ouLastGameRecord),
            (game.awayTeam.ouFavDogOverPct, game.awayTeam.ouFavDogUnderPct, game.awayTeam.ouFavDogRecord,
             game.homeTeam.ouFavDogOverPct, game.homeTeam.ouFavDogUnderPct, game.homeTeam.ouFavDogRecord),
            (game.awayTeam.ouSideFavDogOverPct, game.awayTeam.ouSideFavDogUnderPct, game.awayTeam.ouSideFavDogRecord,
             game.homeTeam.ouSideFavDogOverPct, game.homeTeam.ouSideFavDogUnderPct, game.homeTeam.ouSideFavDogRecord),
            (game.awayTeam.ouRestBucketOverPct, game.awayTeam.ouRestBucketUnderPct, game.awayTeam.ouRestBucketRecord,
             game.homeTeam.ouRestBucketOverPct, game.homeTeam.ouRestBucketUnderPct, game.homeTeam.ouRestBucketRecord),
            (game.awayTeam.ouRestCompOverPct, game.awayTeam.ouRestCompUnderPct, game.awayTeam.ouRestCompRecord,
             game.homeTeam.ouRestCompOverPct, game.homeTeam.ouRestCompUnderPct, game.homeTeam.ouRestCompRecord)
        ]
        for b in buckets {
            let bothOver = (b.awayOver ?? 0) > minPercentage && (b.homeOver ?? 0) > minPercentage
            let bothUnder = (b.awayUnder ?? 0) > minPercentage && (b.homeUnder ?? 0) > minPercentage
            if bothOver {
                let aGames = parseNBARecord(b.awayRec).total
                let hGames = parseNBARecord(b.homeRec).total
                if aGames >= minGamesThreshold && hGames >= minGamesThreshold {
                    let totalGames = Double(aGames + hGames)
                    let avgPct = ((b.awayOver ?? 0) * Double(aGames) + (b.homeOver ?? 0) * Double(hGames)) / totalGames
                    total += avgPct * Double(min(aGames, hGames))
                }
            }
            if bothUnder {
                let aGames = parseNBARecord(b.awayRec).total
                let hGames = parseNBARecord(b.homeRec).total
                if aGames >= minGamesThreshold && hGames >= minGamesThreshold {
                    let totalGames = Double(aGames + hGames)
                    let avgPct = ((b.awayUnder ?? 0) * Double(aGames) + (b.homeUnder ?? 0) * Double(hGames)) / totalGames
                    total += avgPct * Double(min(aGames, hGames))
                }
            }
        }
        return total
    }

    private func atsDominance(for game: NBAGameTrendsData) -> Double {
        var total: Double = 0
        let buckets: [(awayPct: Double?, awayRec: String?, homePct: Double?, homeRec: String?)] = [
            (game.awayTeam.atsLastGameCoverPct, game.awayTeam.atsLastGameRecord,
             game.homeTeam.atsLastGameCoverPct, game.homeTeam.atsLastGameRecord),
            (game.awayTeam.atsFavDogCoverPct, game.awayTeam.atsFavDogRecord,
             game.homeTeam.atsFavDogCoverPct, game.homeTeam.atsFavDogRecord),
            (game.awayTeam.atsSideFavDogCoverPct, game.awayTeam.atsSideFavDogRecord,
             game.homeTeam.atsSideFavDogCoverPct, game.homeTeam.atsSideFavDogRecord),
            (game.awayTeam.atsRestBucketCoverPct, game.awayTeam.atsRestBucketRecord,
             game.homeTeam.atsRestBucketCoverPct, game.homeTeam.atsRestBucketRecord),
            (game.awayTeam.atsRestCompCoverPct, game.awayTeam.atsRestCompRecord,
             game.homeTeam.atsRestCompCoverPct, game.homeTeam.atsRestCompRecord)
        ]
        for b in buckets {
            guard let aPct = b.awayPct, let hPct = b.homePct else { continue }
            let aGames = parseNBARecord(b.awayRec).total
            let hGames = parseNBARecord(b.homeRec).total
            let minGames = min(aGames, hGames)
            if minGames >= minGamesThreshold {
                let diff = abs(aPct - hPct)
                if diff > minATSDifference {
                    total += diff * Double(minGames)
                }
            }
        }
        return total
    }

    private func sortGames(_ list: [NBAGameTrendsData], mode: SortMode) -> [NBAGameTrendsData] {
        switch mode {
        case .ouConsensus:
            return list.sorted { ouConsensusStrength(for: $0) > ouConsensusStrength(for: $1) }
        case .atsDominance:
            return list.sorted { atsDominance(for: $0) > atsDominance(for: $1) }
        case .time:
            return list.sorted { a, b in
                if let aT = a.tipoffTime, let bT = b.tipoffTime { return aT < bT }
                if a.tipoffTime != nil && b.tipoffTime == nil { return true }
                if a.tipoffTime == nil && b.tipoffTime != nil { return false }
                return a.gameDate < b.gameDate
            }
        }
    }

    private struct TipoffRow: Decodable, Sendable {
        let gameId: Int
        let tipoffTimeEt: String?
        enum CodingKeys: String, CodingKey {
            case gameId = "game_id"
            case tipoffTimeEt = "tipoff_time_et"
        }
    }

    // MARK: - Debug helpers

    #if DEBUG
    public func debugSet(_ list: [NBAGameTrendsData], state: LoadState = .loaded) {
        self.games = list
        self.loadState = state
        self.lastFetched = Date()
    }
    #endif
}

File: wagerproof-android-native/core/stores/src/main/kotlin/com/wagerproof/core/stores/NBABettingTrendsStore.kt
package com.wagerproof.core.stores

import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.wagerproof.core.models.NBAGameTrendsData
import com.wagerproof.core.models.NBASituationalTrendRow
import com.wagerproof.core.models.parseNBARecord
import com.wagerproof.core.services.BuildFlags
import com.wagerproof.core.services.SupabaseClients
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.math.abs
import kotlin.math.min

/**
 * Port of iOS `NBABettingTrendsStore.swift` (doc §12.1).
 *
 * NBA situational betting trends. Fetches from
 * `nba_game_situational_trends_today`; falls back to
 * `nba_game_situational_trends` when the today view is empty (matches RN).
 * After loading rows it pairs them by `team_side`, joins tipoff times from
 * `nba_input_values_view`, and computes the O/U consensus + ATS dominance
 * scores used for sorting.
 *
 * Backend queries are byte-identical to RN — select clauses, table names,
 * and post-fetch joins must NOT diverge. NBA/NCAAB tables live on the
 * CFB/sports-data project → `SupabaseClients.cfb`.
 */
@Stable
class NBABettingTrendsStore {
    enum class SortMode { time, ouConsensus, atsDominance }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    var games by mutableStateOf<List<NBAGameTrendsData>>(emptyList()); private set
    var loadState by mutableStateOf<LoadState>(LoadState.Idle); private set

    /**
     * Set on every successful refresh. Drives both the TTL guard below and
     * the game sheets' first-hydrate skeleton rule (`loading && lastFetched
     * == null`). Epoch millis (matches AuthStore's `System.currentTimeMillis`).
     */
    var lastFetched by mutableStateOf<Long?>(null); private set

    private val minGamesThreshold: Int = 5
    private val minPercentage: Double = 55.0
    private val minATSDifference: Double = 10.0

    fun close() = scope.cancel()

    /** Lookup a single game by id. Mirrors RN's `useNBABettingTrendsForGame`. */
    fun trends(forGameId: Int): NBAGameTrendsData? =
        games.firstOrNull { it.gameId == forGameId }

    /**
     * Idempotent hydrate for sheet-local stores — skips the network while a
     * fetch is in flight or the slate is fresh. Mirrors
     * `MLBBettingTrendsStore.refreshIfNeeded`.
     */
    suspend fun refreshIfNeeded(maxAgeSeconds: Long = 600) {
        if (loadState is LoadState.Loading || loadState is LoadState.Refreshing) return
        lastFetched?.let {
            if (System.currentTimeMillis() - it < maxAgeSeconds * 1000L) return
        }
        refresh()
    }

    suspend fun refresh() {
        // FIDELITY-WAIVER #101: iOS DummyDataMode branch (synthesized offseason
        // trends) has no Android DummyData equivalent — skip straight to live.
        loadState = LoadState.Loading
        val cfb = SupabaseClients.cfb

        // Step 1: situational trends (today's table → fallback).
        val trendsRows: List<NBASituationalTrendRow> = try {
            val primary = runCatching {
                cfb.from("nba_game_situational_trends_today")
                    .select {
                        order("game_date", Order.ASCENDING)
                        order("game_id", Order.ASCENDING)
                    }
                    .decodeList<NBASituationalTrendRow>()
            }.getOrDefault(emptyList())
            if (primary.isNotEmpty()) {
                primary
            } else {
                // Fallback path mirrors RN.
                cfb.from("nba_game_situational_trends")
                    .select {
                        order("game_date", Order.ASCENDING)
                        order("game_id", Order.ASCENDING)
                    }
                    .decodeList<NBASituationalTrendRow>()
            }
        } catch (e: Exception) {
            loadState = LoadState.Failed("Failed to fetch NBA trends: ${e.message}")
            return
        }
        if (trendsRows.isEmpty()) {
            games = emptyList()
            loadState = LoadState.Loaded
            lastFetched = System.currentTimeMillis()
            return
        }

        // Step 2: group rows by game_id using team_side.
        val partial = LinkedHashMap<Int, Partial>()
        for (row in trendsRows) {
            val side = row.teamSide.lowercase()
            if (side != "away" && side != "home") continue
            val existing = partial.getOrPut(row.gameId) { Partial(gameDate = row.gameDate) }
            if (side == "away") existing.away = row else existing.home = row
        }

        val built: List<NBAGameTrendsData> = partial.mapNotNull { (gameId, slots) ->
            val away = slots.away ?: return@mapNotNull null
            val home = slots.home ?: return@mapNotNull null
            NBAGameTrendsData(
                gameId = gameId,
                gameDate = slots.gameDate,
                tipoffTime = null,
                awayTeam = away,
                homeTeam = home,
            )
        }

        // Step 3: tipoff times from nba_input_values_view for today's game ids.
        if (built.isNotEmpty()) {
            val gameIds = built.map { it.gameId }
            val timeRows = runCatching {
                cfb.from("nba_input_values_view")
                    .select(Columns.raw("game_id, tipoff_time_et")) {
                        filter { isIn("game_id", gameIds) }
                    }
                    .decodeList<TipoffRow>()
            }.getOrDefault(emptyList())
            val timesMap = HashMap<Int, String>()
            for (row in timeRows) { row.tipoffTimeEt?.let { timesMap[row.gameId] = it } }
            built.forEach { it.tipoffTime = timesMap[it.gameId] }
        }

        games = sortGames(built, SortMode.time)
        loadState = LoadState.Loaded
        lastFetched = System.currentTimeMillis()
    }

    // MARK: - Sort scoring (mirrors RN logic byte-for-byte)

    private fun ouConsensusStrength(game: NBAGameTrendsData): Double {
        var total = 0.0
        val a = game.awayTeam
        val h = game.homeTeam
        val buckets = listOf(
            OuBucket(a.ouLastGameOverPct, a.ouLastGameUnderPct, a.ouLastGameRecord,
                h.ouLastGameOverPct, h.ouLastGameUnderPct, h.ouLastGameRecord),
            OuBucket(a.ouFavDogOverPct, a.ouFavDogUnderPct, a.ouFavDogRecord,
                h.ouFavDogOverPct, h.ouFavDogUnderPct, h.ouFavDogRecord),
            OuBucket(a.ouSideFavDogOverPct, a.ouSideFavDogUnderPct, a.ouSideFavDogRecord,
                h.ouSideFavDogOverPct, h.ouSideFavDogUnderPct, h.ouSideFavDogRecord),
            OuBucket(a.ouRestBucketOverPct, a.ouRestBucketUnderPct, a.ouRestBucketRecord,
                h.ouRestBucketOverPct, h.ouRestBucketUnderPct, h.ouRestBucketRecord),
            OuBucket(a.ouRestCompOverPct, a.ouRestCompUnderPct, a.ouRestCompRecord,
                h.ouRestCompOverPct, h.ouRestCompUnderPct, h.ouRestCompRecord),
        )
        for (b in buckets) {
            val bothOver = (b.awayOver ?: 0.0) > minPercentage && (b.homeOver ?: 0.0) > minPercentage
            val bothUnder = (b.awayUnder ?: 0.0) > minPercentage && (b.homeUnder ?: 0.0) > minPercentage
            if (bothOver) {
                val aGames = parseNBARecord(b.awayRec).total
                val hGames = parseNBARecord(b.homeRec).total
                if (aGames >= minGamesThreshold && hGames >= minGamesThreshold) {
                    val totalGames = (aGames + hGames).toDouble()
                    val avgPct = ((b.awayOver ?: 0.0) * aGames + (b.homeOver ?: 0.0) * hGames) / totalGames
                    total += avgPct * min(aGames, hGames)
                }
            }
            if (bothUnder) {
                val aGames = parseNBARecord(b.awayRec).total
                val hGames = parseNBARecord(b.homeRec).total
                if (aGames >= minGamesThreshold && hGames >= minGamesThreshold) {
                    val totalGames = (aGames + hGames).toDouble()
                    val avgPct = ((b.awayUnder ?: 0.0) * aGames + (b.homeUnder ?: 0.0) * hGames) / totalGames
                    total += avgPct * min(aGames, hGames)
                }
            }
        }
        return total
    }

    private fun atsDominance(game: NBAGameTrendsData): Double {
        var total = 0.0
        val a = game.awayTeam
        val h = game.homeTeam
        val buckets = listOf(
            AtsBucket(a.atsLastGameCoverPct, a.atsLastGameRecord, h.atsLastGameCoverPct, h.atsLastGameRecord),
            AtsBucket(a.atsFavDogCoverPct, a.atsFavDogRecord, h.atsFavDogCoverPct, h.atsFavDogRecord),
            AtsBucket(a.atsSideFavDogCoverPct, a.atsSideFavDogRecord, h.atsSideFavDogCoverPct, h.atsSideFavDogRecord),
            AtsBucket(a.atsRestBucketCoverPct, a.atsRestBucketRecord, h.atsRestBucketCoverPct, h.atsRestBucketRecord),
            AtsBucket(a.atsRestCompCoverPct, a.atsRestCompRecord, h.atsRestCompCoverPct, h.atsRestCompRecord),
        )
        for (b in buckets) {
            val aPct = b.awayPct ?: continue
            val hPct = b.homePct ?: continue
            val aGames = parseNBARecord(b.awayRec).total
            val hGames = parseNBARecord(b.homeRec).total
            val minGames = min(aGames, hGames)
            if (minGames >= minGamesThreshold) {
                val diff = abs(aPct - hPct)
                if (diff > minATSDifference) {
                    total += diff * minGames
                }
            }
        }
        return total
    }

    private fun sortGames(list: List<NBAGameTrendsData>, mode: SortMode): List<NBAGameTrendsData> =
        when (mode) {
            // NBA computes scores on demand at sort time (NOT cached on the model,
            // unlike the NCAAB/MLB siblings).
            SortMode.ouConsensus -> list.sortedByDescending { ouConsensusStrength(it) }
            SortMode.atsDominance -> list.sortedByDescending { atsDominance(it) }
            SortMode.time -> list.sortedWith(TIME_COMPARATOR)
        }

    private class Partial(
        var away: NBASituationalTrendRow? = null,
        var home: NBASituationalTrendRow? = null,
        val gameDate: String,
    )

    private data class OuBucket(
        val awayOver: Double?, val awayUnder: Double?, val awayRec: String?,
        val homeOver: Double?, val homeUnder: Double?, val homeRec: String?,
    )

    private data class AtsBucket(
        val awayPct: Double?, val awayRec: String?, val homePct: Double?, val homeRec: String?,
    )

    @Serializable
    private data class TipoffRow(
        @SerialName("game_id") val gameId: Int,
        @SerialName("tipoff_time_et") val tipoffTimeEt: String? = null,
    )

    companion object {
        /** some-tipoff-before-none, else compare tipoff string, else gameDate. */
        private val TIME_COMPARATOR = Comparator<NBAGameTrendsData> { a, b ->
            val aT = a.tipoffTime
            val bT = b.tipoffTime
            when {
                aT != null && bT != null -> aT.compareTo(bT)
                aT != null && bT == null -> -1
                aT == null && bT != null -> 1
                else -> a.gameDate.compareTo(b.gameDate)
            }
        }
    }
}
